nixos/curator: init elasticsearch curator

shmish111 opened the pull request:
See [elasticsearch curator](https://www.elastic.co/guide/en/elasticsearch/client/curator/5.5/index.html)

###### Motivation for this change

I needed to use this tool on my NixOS ES cluster

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

shmish111 committed:
nixos/curator: init elasticsearch curator

https://www.elastic.co/guide/en/elasticsearch/client/curator/5.5/index.html

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,93 @@
+
+  options.services.curator = {
+
+    enable = mkOption {
```
Maybe use [mkEnableOption](https://github.com/NixOS/nixpkgs/blob/master/lib/options.nix#L27) here.

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,93 @@
+    };
+in {
+
+  options.services.curator = {
```
I think we should name the module `elasticsearch-curator` instead of just `curator` because it's more descriptive and consistent with the name of the Python package.

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,93 @@
+
+  config = mkIf cfg.enable {
+
+    systemd.services.curator = {
```
As with the module name I think we should also name the systemd unit `elasticsearch-curator` instead of just `curator`.

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,93 @@
+      enable = cfg.enable;
+      startAt = cfg.interval;
+      serviceConfig = {
+      ExecStart = ''${pkgs.python36Packages.elasticsearch-curator}/bin/curator --config ${curatorConfig} ${curatorAction}'';
```
\<nitpicking> could you indent this line? \</nitpicking>

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,93 @@
+        # not a Python "NoneType"
+        client:
+          hosts: ${builtins.toJSON cfg.hosts}
+          port: 9200
```
Shouldn't the port also be configurable? We should probably default it to [services.elasticsearch.port](https://github.com/NixOS/nixpkgs/blob/72f3a5cf5ce77ac8c4d431f104e06fe94cdf7366/nixos/modules/services/search/elasticsearch.nix#L72).

shmish111 commented:
@basvandijk I have updated the PR according to your PR, do I need to squash the extra commit?

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,94 @@
+with lib;
+
+let
+    cfg = config.services.curator;
```
This should now be: `config.services.elasticsearch-curator`

basvandijk commented on nixos/modules/services/search/curator.nix:
```diff
@@ -0,0 +1,94 @@
+    };
+    port = mkOption {
+      description = "the port that elasticsearch is listening on";
+      type = types.int
```
You're missing a semicolon here.

basvandijk reviewed (changes requested):
Could you also rename the file to `elasticsearch-curator.nix`?

shmish111 commented:
fixed. should I squash @basvandijk ?

basvandijk commented:
@shmish111 thank you. A squash would be nice.

Any chance you could add a NixOS test for this? Maybe extend the [existing ELK test](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/elk.nix).

shmish111 commented:
@basvandijk do you think it is ok to add it to the ELK test? I feel like it would be unnecessary overhead to have a test that boots up another ES machine when I could just add in some test cases to the existing ELK test. I suppose I would just create an index in ES, run a curator action that should delete it and then check it has been deleted.

basvandijk commented:
@shmish111 extending the existing ELK test like you suggest is the way to go IMHO. It would not only test curator but also elasticsearch itself.

shmish111 commented:
@basvandijk how do I run the nixpkgs tests? I can't find documentation anywhere :-(

shmish111 commented:
@basvandijk I have squashed. Also, I can't get the test passing on `master` so I can't really do anything until I do that. 

```
uname -a
Linux ip-10-0-4-156.us-east-2.compute.internal 4.14.32 #1-NixOS SMP Sat Mar 31 16:10:43 UTC 2018 x86_64 GNU/Linux

nix-build nixos/tests/elk.nix
```

This fails to connect to elasticsearch. It seems that elasticsearch has failed to start but I can't work out how to debug in order to find out why.

basvandijk commented:
See the following on how to run tests interactively:
https://nixos.org/nixos/manual/index.html#sec-running-nixos-tests
That will get you into a console on the VM with which you can debug the issue.

shmish111 commented:
@basvandijk I've tried to play around but it's difficult to debug as I can't get into a shell in the running VM. All I can see is the output of `startAll` etc. I managed to see that elasticsearch does start eventually however it take a long long time for some reason. Then logstash is failing, with `"TypeError('Invalid argument: family must be 4 or 6');"`. Also `cat /tmp/logstash.out` never appears.

Have you managed to run these tests successfully? I am trying to run these on a NixOS machine on AWS with the image `NixOS x86_64-linux 18.03.132946.1caae7247b8 (hvm-ebs)` as a `t2.xlarge` instance.

I think the test maintainers need to look at this, does this need to prevent my PR? I'm happy to add a test but the existing test needs to be fixed first and at the moment that is beyond me.

basvandijk commented on nixos/modules/services/search/elasticsearch-curator.nix:
```diff
@@ -0,0 +1,94 @@
+  config = mkIf cfg.enable {
+
+    systemd.services.elasticsearch-curator = {
+      enable = cfg.enable;
```
You don't to set this since `enable` is `true` by default and the whole definition is already under a `mkIf`.

basvandijk reviewed (changes requested):
Regarding the test, I can successfully run `nix-build nixos/tests/elk.nix` on my Mac Book Pro mid 2015 running NixOS 18.03. Maybe a `t2.xlarge` can't run VMs?

Do you have some statements available that create an index on elasticsearch and an `actionYAML` definition that deletes that index? Then I can try running that test to see if it works.

shmish111 committed:
elasticsearch-curator: don't need to add enable to elasticsearch-curator service

shmish111 committed:
elasticsearch-curator: add test

shmish111 commented:
@basvandijk ok, I've fixed the `cfg.enable` thing. I've also added a test for curator, this just runs curator after the other tests and checks that the logstash index has been deleted. This index must have been created for the logstash queries to succeed. Can you run these tests because I have no access to a NixOS machine that isn't on AWS at the moment, and I suspect you are correct and that VMs aren't running correctly on AWS instances.

basvandijk committed:
elasticsearch-curator: include the module in the module-list & fix bug

basvandijk committed:
elasticsearch-curator: include missing boto3 dependency

TODO: we're still missing the requests-aws4auth>=0.9 dependency

basvandijk commented:
Great we're getting closer! I pushed some more fixes to your branch. In the end you might want to squash all of them.

Could you look into fixing the `elasticsearch-curator` python package because building it fails with:

```
nix-build -A python36Packages.elasticsearch-curator
...
Collecting requests-aws4auth>=0.9 (from elasticsearch-curator==5.5.4)
  Could not find a version that satisfies the requirement requests-aws4auth>=0.9 (from elasticsearch-curator==5.5.4) (from versions: )
No matching distribution found for requests-aws4auth>=0.9 (from elasticsearch-curator==5.5.4)
```

I see we don't have a package for `requests-aws4auth` so that needs to be packaged.

basvandijk commented:
You might also want to add yourself as maintainer of the module using:

```
  meta.maintainers = with maintainers; [ shmish111  ];
```

and add `shmish111` to the [maintainer-list.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix).

basvandijk committed:
pythonPackages.requests-aws4auth: init at 0.9

Needed by elasticsearch-curator.

basvandijk committed:
elasticsearch-curator: disable test-suite because it hangs

basvandijk committed:
nixos/tests/elk.nix: make sure the test doesn't wait for too long on elasticsearch-curator

basvandijk commented:
@GrahamcOfBorg test elk.ELK-5 elk.ELK-6

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44340&attempt_id=f4cdbe04-d1ba-42de-be9b-5b0a3eb1d90c)

The following builds were skipped because they don't evaluate on aarch64-linux: tests.elk.ELK-5, tests.elk.ELK-6

<details><summary>Partial log (click to expand)</summary><p>

```
Cannot nix-instantiate `tests.elk.ELK-5' because:
error: attribute 'ELK-5' in selection path 'tests.elk.ELK-5' not found

Cannot nix-instantiate `tests.elk.ELK-6' because:
error: attribute 'ELK-6' in selection path 'tests.elk.ELK-6' not found

```
</p></details>

GrahamcOfBorg commented:
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44340&attempt_id=0ab8dd9a-9dc8-43fc-a0c1-2789ac839dea)

Attempted: tests.elk.ELK-5, tests.elk.ELK-6

<details><summary>Partial log (click to expand)</summary><p>

```
one: running command: cat /tmp/logstash.out | grep flowers
one# cat: /tmp/logstash.out: No such file or directory
one: exit status 1
one: running command: cat /tmp/logstash.out | grep flowers
one# cat: /tmp/logstash.out: No such file or directory
one: exit status 1
one: running command: cat /tmp/logstash.out | grep flowers
one# cat: /tmp/logstash.out: No such file or directory
building of '/nix/store/q5wv95vqb8bifpzyz0qvcvb78r5vddi8-vm-test-run-ELK-5.drv' timed out after 1800 seconds
error: build of '/nix/store/mp71gw6kysf1y7nilnnvz5gq70dyh7di-vm-test-run-ELK-6.drv', '/nix/store/q5wv95vqb8bifpzyz0qvcvb78r5vddi8-vm-test-run-ELK-5.drv' failed
```
</p></details>

FRidh commented on pkgs/development/python-modules/requests-aws4auth/default.nix:
```diff
@@ -0,0 +1,21 @@
+
+  propagatedBuildInputs = [ requests ];
+
+  doCheck = false;
```
Why are the tests disabled? Please include a comment.

basvandijk replied:
The test hang. I'm not sure why. Good point about adding a comment. Will do.

basvandijk committed:
elasticsearch-curator: add note to the NixOS release notes

basvandijk committed:
requests-aws4auth: only disable tests on python >= 3

basvandijk committed:
elasticsearch-curator: explain why we disable the test-suite
